File: packages/python/src/output_guard/strategies/fix_closers.py
"""fix-closers: Append missing closing braces/brackets to balance the structure.

Mirror of packages/ts/src/strategies/fix-closers.ts.
"""

from __future__ import annotations


def fix_closers(text: str) -> str:
    stack: list[str] = []
    in_string = False
    escaped = False
    for ch in text:
        if escaped:
            escaped = False
            continue
        if ch == "\\" and in_string:
            escaped = True
            continue
        if ch == '"':
            in_string = not in_string
            continue
        if in_string:
            continue
        if ch == "{":
            stack.append("}")
        elif ch == "[":
            stack.append("]")
        elif ch in ("}", "]") and stack and stack[-1] == ch:
            stack.pop()
    return text + "".join(reversed(stack))

File: packages/ts/src/strategies/fix-closers.ts
/**
 * fix-closers: Append missing closing braces/brackets to balance the structure.
 * Must-isolate: runs after fix-truncated and fix-ellipsis.
 */
export const fixClosers = (text: string): string => {
  const stack: string[] = [];
  let inString = false;
  let escaped = false;

  for (const ch of text) {
    if (escaped) { escaped = false; continue; }
    if (ch === "\\" && inString) { escaped = true; continue; }
    if (ch === '"') { inString = !inString; continue; }
    if (inString) continue;
    if (ch === "{") stack.push("}");
    else if (ch === "[") stack.push("]");
    else if (ch === "}" || ch === "]") {
      if (stack[stack.length - 1] === ch) stack.pop();
    }
  }

  // Append all missing closers in reverse
  return text + stack.reverse().join("");
};
